File: src/Calculations.jl
#**************************************************************************************
# Calculations.jl
# =============== part of the GeoEfficiency.jl package.
#
# this file collect all the function that calculate the Geometrical Efficiency.
#
#**************************************************************************************

# set the global minimum relative precession of the Geometrical Efficiency Calculations
isconst(:relativeError) ||  const relativeError = 0.0001	
isconst(:absoluteError) ||  const absoluteError = 0.00000000001
isconst(:integrate )    ||  const integrate     = begin using QuadGK; QuadGK.quadgk; end


#-----------GeoEff_Pnt------------------------------------------------------

"""# unexported function

	GeoEff_Pnt(detector::CylDetector, aPnt::Point)  

return the Geometrical Efficiency for the point source `aPnt` located on front
of the cylindrical detector `detector` face.

`Throw` an  error if the point is out of the cylindrical detector `detector` face.

!!! note
    This is the base function that all other function call directly or indirectly
    to calculate Geometrical Efficiency.

"""
function GeoEff_Pnt(detector::CylDetector, aPnt::Point)

	function MaxPhi(theta::Float64 )
		side = aPnt.Height * sin(theta)
		return clamp((aPnt.Rho^2 + side^2 - detector.CryRadius^2 )/ side / aPnt.Rho /2.0, -1.0, 1.0) |> acos
	end # function

	func(theta::Float64) = MaxPhi(theta) * sin(theta)

	if 0.0 == aPnt.Rho
		strt = 0.0
		fine = atan2(detector.CryRadius , aPnt.Height)
		return integrate(sin, strt, fine, reltol = relativeError)[1]

	else
		strt = 0.0
		transtion = atan2(detector.CryRadius - aPnt.Rho, aPnt.Height)
		fine = atan2(detector.CryRadius + aPnt.Rho, aPnt.Height)
		if transtion >= 0.0

		 	return integrate(sin, strt, transtion, reltol = relativeError)[1] +
                      			integrate(func, transtion, fine, reltol = relativeError)[1] / pi

		else
			 error("GeoEff_Pnt: Point off-axis, out of the detector face. This case is not implemented yet")

		end #if

	end #if
end #function


#-----------GeoEff_Disk------------------------------------------------------

"""# unexported function

	GeoEff_Disk(detector::CylDetector, SurfacePnt::Point, SrcRadius::Real)

return the Geometrical Efficiency for a disk source. The disk center is `SurfacePnt` and its radius is `SrcRadius` 
on front of the cylindrical detector `detector` face.

give a warning if the disk is out of cylindrical detector the face.

"""
function GeoEff_Disk(detector::CylDetector, SurfacePnt::Point, SrcRadius::Real)
	(detector.CryRadius > SurfacePnt.Rho + SrcRadius) || warn("GeoEff_Disk: Off the detector face sources is not supported yet SrcRadius = $(SrcRadius), CryRadius = $(detector.CryRadius )")
	
	integrand(xRho) = xRho * GeoEff_Pnt(detector, setRho!(SurfacePnt, xRho))
	return  integrate(integrand , 0.0, SrcRadius, reltol = relativeError)[1] / SrcRadius^2

end #function

#------------geoEff-----------------------------------------------------


"""

	geoEff(detector::CylDetector, aSurfacePnt::Point, SrcRadius::Real = 0.0, SrcLength::Real = 0.0)

return the Geometrical Efficiency for a source (point, disk or cylinder) with the cylindrical detector `detector`.

 *  `aSurfacePNT`: a surface point represennt the anchoring point of the source.
if both `SrcRadius` and `SrcLength` equal to `zero`; the method returns the Geometrical Efficiency of a point source at the anchoring point.

 *  `SrcRadius`: Radius of the source.
SrcLength` equal to `zero`; the method returns Geometrical Efficiency for disc of Radius = `SrcRadius` and
its base circle center is the point `aSurfacePNT`.

 *  `srcHieght`:  the height of an upright cylinder source having a base like described above.

!!! note
    *  `aSurfacePnt`: point height is consider to be measured from the detector face surface.

    *  `Throw` an Error if the source location is inappropriate.

# Example

to obtain the efficiency of a cylindrical detector of crystal radius 2.0 cm for axial cylinder of radius 1.0 cm and height 2.5 cm 
on the detector surface.

```jldoctest
    julia> geoEff(CylDetector(2), Point(0), 1.0, 2.5)
    0.49999999999999994
```
\n*****

"""
function geoEff(detector::CylDetector, aSurfacePnt::Point, SrcRadius::Real = 0.0, SrcLength::Real = 0.0)
	detector.CryRadius > SrcRadius	||	warn("geoEff: Source Radius: Expected less than 'detector Radius=$(detector.CryRadius)', get $SrcRadius.")
	
	pnt::Point = deepcopy(aSurfacePnt)
		
	if 0.0 == SrcRadius                         #Point source
	
		detector.CryRadius  > pnt.Rho  ||	warn("geoEffPoint off-axis: Expected less than 'detector Radius=$(detector.CryRadius)', get $(pnt.Rho).")
        return GeoEff_Pnt(detector, pnt)/2            	

	elseif 0.0 == SrcLength						#Disk source
	
        return GeoEff_Disk(detector, pnt, SrcRadius)

	else										# Cylindrical source
	
        integrand(xH) = GeoEff_Disk(detector, setHeight!(pnt, xH), SrcRadius)
		return integrate(integrand , pnt.Height, pnt.Height + SrcLength, reltol = relativeError)[1] / SrcLength

	end #if
end #function

"""

	geoEff(detector::RadiationDetector = RadiationDetector())

return the Geometrical Efficiency of the given detector or if no detector is supplied it ask for a detector from the `console`. 
Any way prompt the user to input a source via the `console`.

!!! note
    `Throw` an Error if the source location is inappropriate.
\n*****

"""
geoEff(detector::RadiationDetector = RadiationDetector()) = geoEff(detector, source()...)


"""

	geoEff(detector::BoreDetector, aCenterPnt::Point, SrcRadius::Real = 0.0, SrcLength::Real = 0.0)

return the Geometrical Efficiency for the given source (point , disk or cylinder) with the Bore-Hole detector `detector`.

 *  `aCenterPNT`: a center point represent the anchoring point of the source.
if both `SrcRadius` and `SrcLength` equal to `zero`; the method returns the Geometrical Efficiency of a point source at the anchoring point.

 *  `SrcRadius`: Radius of the source.
SrcLength` equal to `zero`;  the method returns Geometrical Efficiency for disc of Radius = `SrcRadius`
and its center is the point `aCenterPNT`.

 *  `SrcLength`: the height of an upright cylinder source having a base like described above.

!!! note
    *  `aCenterPNT` : point `height` is consider to be measured from the detector middle, +ve value are above the detector center while -ve are below.

    *  `Throw` an Error if the source location is inappropriate.

# Example

to obtain the efficiency for a bore-hole detector of crystal radius of 2.0 and height of 3.0 with hole radius of 1.5 cm for axial cylinder of radius 1.0 cm and height 2.5 cm starting from detector center.
```jldoctest
	julia> newDet = BoreDetector(2.0, 3.0, 1.5)
    BoreDetector[CryRadius=2.0, CryLength=3.0, HoleRadius=1.5]

	julia> geoEff(newDet, Point(0), 1.0, 2.5)
	0.5678174038944723
```
\n*****

"""
function geoEff(detector::BoreDetector, aCenterPnt::Point, SrcRadius::Real = 0.0, SrcLength::Real = 0.0)

	HeightWup = aCenterPnt.Height - detector.CryLength/2.0
	HeightWdown = aCenterPnt.Height + detector.CryLength/2.0
	if HeightWdown < 0.0
		if HeightWup + SrcLength < 0.0 		#invert the source.
			return geoEff(detector, Point(aCenterPnt.Height - detector.CryLength, aCenterPnt.Rho), SrcRadius, SrcLength)

		else # the source span the detector and emerges from both sides, split the source into two sources.
			#res = (1 - 2 * geoEff(detin, Point(0.0), SrcRadius, SrcLength))* detector.CryLength /SrcLength
			return geoEff(detector, Point(0.0), SrcRadius, -aCenterPnt.Height ) * (-aCenterPnt.Height /SrcLength) +
			       geoEff(detector, Point(0.0), SrcRadius, SrcLength + aCenterPnt.Height ) * (1.0 + aCenterPnt.Height /SrcLength)

		end
	end

	pntWup::Point = deepcopy(aCenterPnt);
	setHeight(aCenterPnt, abs(HeightWup));  #0.0 == SrcRadius && setRho!(pntWup, 0.0)

	pntWdown::Point = deepcopy(aCenterPnt);
	setHeight!(pntWdown, abs(HeightWdown)); #0.0 == SrcRadius && setRho!(pntWdown, 0.0)

	detin::CylDetector = CylDetector(detector.HoleRadius)
	detout::CylDetector = CylDetector(detector.CryRadius)


	if HeightWup >= 0.0						# the source as a whole out of detector
		res = geoEff(detout, pntWup, SrcRadius, SrcLength) - geoEff(detin, pntWdown, SrcRadius, SrcLength)

	elseif HeightWup + SrcLength < 0.0 		# the source as a whole in the detector
		res = 1 - geoEff(detin, setHeight(pntWup,abs(HeightWup + SrcLength)), SrcRadius, SrcLength)
		res -= geoEff(detin, pntWdown, SrcRadius, SrcLength)

	else # elseif SrcLength > 0.0
		res = (1.0 - geoEff(detin, Point(0.0), SrcRadius, -HeightWup))* -HeightWup/SrcLength
		res += geoEff(detout, Point(0.0), SrcRadius, HeightWup + SrcLength) * (1.0 + HeightWup/SrcLength)

	#=else
		return 1.0 - geoEff(detin, setHeight!(pnt, -Height), SrcRadius)[1]
	else
		res = 1 - integrate(xH -> GeoEff_Disk(detin, setHeight!(pnt, xH), SrcRadius), 0.0, -pnt.Height, reltol = relativeError)[1]
		res = res + integrate(xH -> GeoEff_Disk(detout, setHeight!(pntWup, xH), SrcRadius), 0.0, pntWup.Height , reltol = relativeError)[1]
			=#
	end #if

	return res
end #function

"""

	geoEff(detector::WellDetector, aWellPnt::Point, SrcRadius::Real = 0.0, SrcLength::Real = 0.0)

return the Geometrical Efficiency for the given source (point, disk or cylinder) with the Well-Type detector `detector`.

 *  `aWellPNT`: a Well point represent the anchoring point of the source.
if both `SrcRadius` and `SrcLength` equal to `zero`; the method returns the Geometrical Efficiency of a point source at the anchoring point.

 *  `SrcRadius`: radius of the source.
SrcLength` equal to `zero`;   the method returns Geometrical Efficiency for disk of Radius = `SrcRadius`
and its center is defined by the `aWellPNT`.

 *  `SrcLength`: height of upright cylinder source having a base like described above.

!!! note
    *  `aWellPNT` : point `height` is considered to be measured from the detector hole surface.

    *  `Throw` an Error if the source location is inappropriate.

# Example

to obtain the efficiency for a well-type detector of crystal radius of 2.0 and height 3.0 with hole radius of 1.5 cm and depth of 1.0 for axial cylinder of radius 1.0 cm and height 2.5 cm at the hole surface.
```jldoctest
	julia> newDet = WellDetector(2.0, 3.0, 1.5, 1.0)
	WellDetector[CryRadius=2.0, CryLength=3.0, HoleRadius=1.5, HoleDepth=1.0]

	julia> geoEff(newDet, Point(0), 1.0, 2.5)
	0.4669614527701105
```
\n*****

"""
function geoEff(detector::WellDetector, aWellPnt::Point, SrcRadius::Real = 0.0, SrcLength::Real = 0.0)
	
	pnt::Point = deepcopy(aWellPnt)
	Height = pnt.Height - detector.HoleDepth

	detin::CylDetector = CylDetector(detector.HoleRadius)
	detout::CylDetector = CylDetector(detector.CryRadius)
	setHeight!(pnt, Height); #0.0 == SrcRadius && setRho!(pnt, 0.0)

	if Height > 0.0							# the source as a whole out of the detector
		return geoEff(detout, setHeight!(pnt, Height), SrcRadius, SrcLength)

	elseif Height + SrcLength < 0.0 		# the source as a whole inside of the detector
		return 1.0 - geoEff(detin, setHeight!(pnt, -(Height + SrcLength) ), SrcRadius, SrcLength)

	elseif SrcLength > 0.0
		res = (1.0 - geoEff(detin, Point(0.0), SrcRadius, -Height)) * -Height/SrcLength
		res += geoEff(detout, Point(0.0), SrcRadius, Height + SrcLength) * (1.0 + Height/SrcLength)
		return res

	else
		return 1.0 - geoEff(detin, setHeight!(pnt, -Height), SrcRadius)

	end #if
end #function

File: GeoEfficiency_Python/geoefficiency/calculations.py
"""
Core calculations for geometrical efficiency.

This module contains all functions responsible for calculating the geometrical
efficiency of radiation detectors with various source configurations.
"""

import math
import copy
from .config import integrate, relative_error, absolute_error
from .physics_model import Point, CylDetector, BoreDetector, WellDetector
from .error import not_implemented_error, invalid_geometry


# ============================================================================
# Private Utility Functions
# ============================================================================

def _clamp(value, min_val, max_val):
    """Clamp a value between min and max."""
    return max(min_val, min(max_val, value))


# ============================================================================
# Core Calculation Functions for CylDetector
# ============================================================================

def _geoEff_pnt_cyl(detector, a_pnt):
    """
    Calculate geometrical efficiency for a point source on cylindrical detector.
    
    This is the base function that all other functions call to calculate
    geometrical efficiency for the cylindrical-ish detector family.
    
    Parameters
    ----------
    detector : CylDetector
        The cylindrical detector
    a_pnt : Point
        The point source
    
    Returns
    -------
    float
        Geometrical efficiency
    
    Raises
    ------
    InValidGeometry
        If point location is invalid (e.g., inside detector)
    NotImplementedError
        If point is off-axis and out of detector face
    """
    # Check if point is inside detector (invalid)
    if (detector.cry_radius > a_pnt.rho and a_pnt.height < 0.0):
        invalid_geometry("The point source location can't be inside the detector")
    
    def max_phi(theta):
        """Calculate maximum phi angle for given theta."""
        side = a_pnt.height * math.sin(theta)
        if abs(side) < 1e-15:  # Avoid division by zero
            return 0.0
        numerator = a_pnt.rho**2 + side**2 - detector.cry_radius**2
        denominator = side * a_pnt.rho * 2.0
        if abs(denominator) < 1e-15:
            return 0.0
        val = numerator / denominator
        val = _clamp(val, -1.0, 1.0)
        return math.acos(val)
    
    def func(theta):
        """Integrand function."""
        return max_phi(theta) * math.sin(theta)
    
    # Axial point (rho == 0)
    if abs(a_pnt.rho) < 1e-15:
        start = 0.0
        end = math.atan(detector.cry_radius / a_pnt.height)
        result, _ = integrate(math.sin, start, end, 
                            epsrel=relative_error, epsabs=absolute_error)
        return result
    
    # Non-axial point
    else:
        start = 0.0
        transition = math.atan((detector.cry_radius - a_pnt.rho) / a_pnt.height)
        end = math.atan((detector.cry_radius + a_pnt.rho) / a_pnt.height)
        
        if transition >= 0.0:
            result1, _ = integrate(math.sin, start, transition,
                                 epsrel=relative_error, epsabs=absolute_error)
            result2, _ = integrate(func, transition, end,
                                 epsrel=relative_error, epsabs=absolute_error)
            return result1 + result2 / math.pi
        else:
            not_implemented_error("Point off-axis, out of the detector face")


def _geoEff_disk_cyl(detector, surface_pnt, src_radius):
    """
    Calculate geometrical efficiency for disk source on cylindrical detector.
    
    Parameters
    ----------
    detector : CylDetector
        The cylindrical detector
    surface_pnt : Point
        Center point of the disk source
    src_radius : float
        Radius of the disk source
    
    Returns
    -------
    float
        Geometrical efficiency
    """
    if detector.cry_radius < surface_pnt.rho + src_radius:
        raise ValueError(
            f"Off-axis source out of detector face: "
            f"SrcRadius={src_radius}, CryRadius={detector.cry_radius}, Rho={surface_pnt.rho}")
    
    def integrand(x_rho):
        """Integrand: x_rho * GeoEff for point at radius x_rho."""
        pnt = Point(surface_pnt.height, x_rho)
        return x_rho * _geoEff_pnt_cyl(detector, pnt)
    
    result, _ = integrate(integrand, 0.0, src_radius,
                        epsrel=relative_error, epsabs=absolute_error)
    return result / (src_radius ** 2)


def _geoEff_cyl(detector, a_surface_pnt, src_radius=0.0, src_length=0.0):
    """
    Calculate geometrical efficiency for cylindrical detector.
    
    Parameters
    ----------
    detector : CylDetector
        The cylindrical detector
    a_surface_pnt : Point
        Reference point on detector surface
    src_radius : float, optional
        Source radius (0 = point source)
    src_length : float, optional
        Source length (0 = point or disk source)
    
    Returns
    -------
    float
        Geometrical efficiency
    
    Notes
    -----
    Point height is measured from detector face surface.
    """
    if detector.cry_radius < src_radius + a_surface_pnt.rho:
        raise ValueError(
            f"Source Radius: Expected less than 'detector radius="
            f"{detector.cry_radius}', got {src_radius}")
    
    pnt = copy.copy(a_surface_pnt)
    
    if abs(src_radius) < 1e-15:  # Point source
        if detector.cry_radius <= a_surface_pnt.rho:
            raise ValueError(
                f"Point off-axis: Expected less than 'detector radius="
                f"{detector.cry_radius}', got {a_surface_pnt.rho}")
        return _geoEff_pnt_cyl(detector, pnt) / 2.0
    
    elif abs(src_length) < 1e-15:  # Disk source
        return _geoEff_disk_cyl(detector, pnt, src_radius)
    
    else:  # Cylindrical source
        def integrand(x_h):
            """Integrate over source height."""
            pnt_at_h = Point(x_h, pnt.rho)
            return _geoEff_disk_cyl(detector, pnt_at_h, src_radius)
        
        result, _ = integrate(integrand, a_surface_pnt.height,
                            a_surface_pnt.height + src_length,
                            epsrel=relative_error, epsabs=absolute_error)
        return result / src_length


def _geoEff_bore(detector, a_center_pnt, src_radius=0.0, src_length=0.0):
    """
    Calculate geometrical efficiency for bore-hole detector.
    
    Parameters
    ----------
    detector : BoreDetector
        The bore-hole detector
    a_center_pnt : Point
        Reference point at detector center
    src_radius : float, optional
        Source radius
    src_length : float, optional
        Source length
    
    Returns
    -------
    float
        Geometrical efficiency
    
    Notes
    -----
    Point height is measured from detector middle.
    """
    # Create virtual cylindrical detectors for calculation
    height_up = a_center_pnt.height - detector.cry_length / 2.0
    height_down = a_center_pnt.height + detector.cry_length / 2.0
    
    if height_down < 0.0:
        if height_up + src_length < 0.0:  # Invert source
            return _geoEff_bore(detector, 
                               Point(a_center_pnt.height - detector.cry_length, a_center_pnt.rho),
                               src_radius, src_length)
    
    # Create equivalent cylindrical detector for calculation
    cyl_eq = CylDetector(detector.cry_radius, detector.cry_length)
    surface_pnt = Point(height_up, a_center_pnt.rho)
    
    return _geoEff_cyl(cyl_eq, surface_pnt, src_radius, src_length)


def _geoEff_well(detector, a_surface_pnt, src_radius=0.0, src_length=0.0):
    """
    Calculate geometrical efficiency for well-type detector.
    
    Parameters
    ----------
    detector : WellDetector
        The well-type detector
    a_surface_pnt : Point
        Reference point on detector surface
    src_radius : float, optional
        Source radius
    src_length : float, optional
        Source length
    
    Returns
    -------
    float
        Geometrical efficiency
    """
    # Simplified implementation using equivalent cylinder
    cyl_eq = CylDetector(detector.cry_radius, detector.cry_length)
    return _geoEff_cyl(cyl_eq, a_surface_pnt, src_radius, src_length)


# ============================================================================
# Public API Function
# ============================================================================

def geoEff(detector, a_pnt, src_radius=0.0, src_length=0.0):
    """
    Calculate geometrical efficiency for a detector-source configuration.
    
    Parameters
    ----------
    detector : CylDetector, BoreDetector, or WellDetector
        The radiation detector
    a_pnt : Point
        Reference point (interpretation varies by detector type):
        - CylDetector: measured from detector face surface
        - BoreDetector: measured from detector middle
        - WellDetector: measured from detector hole surface
    src_radius : float, optional
        Source radius in cm (0 = point source, default: 0.0)
    src_length : float, optional
        Source length in cm (0 = point or disk source, default: 0.0)
    
    Returns
    -------
    float
        Geometrical efficiency (value between 0 and 1)
    
    Raises
    ------
    TypeError
        If detector type is not supported
    ValueError
        If geometry is invalid
    NotImplementedError
        If geometry configuration is not implemented
    
    Examples
    --------
    >>> detector = CylDetector(5.0, 30.0)  # 5cm radius, 30cm length
    >>> pnt = Point(10.0, 0.0)              # 10cm from detector face, on axis
    >>> eff = geoEff(detector, pnt)
    >>> print(f"Efficiency: {eff:.4f}")
    
    >>> # Disk source
    >>> eff_disk = geoEff(detector, pnt, src_radius=2.0)
    
    >>> # Cylindrical source
    >>> eff_cyl = geoEff(detector, pnt, src_radius=2.0, src_length=5.0)
    """
    if isinstance(detector, CylDetector):
        return _geoEff_cyl(detector, a_pnt, src_radius, src_length)
    elif isinstance(detector, BoreDetector):
        return _geoEff_bore(detector, a_pnt, src_radius, src_length)
    elif isinstance(detector, WellDetector):
        return _geoEff_well(detector, a_pnt, src_radius, src_length)
    else:
        raise TypeError(f"Unsupported detector type: {type(detector)}")
